crates/sui-graphql-rpc/src/config.rs

```diff
@@ -132,8 +132,54 @@ pub struct BackgroundTasksConfig {
     pub watermark_update_ms: u64,
 }
 
-#[derive(Debug)]
-pub struct Version(pub &'static str);
+/// The Version of the service. major.minor combination represents the major release.
+/// The patch represents a fix in that release.
+#[derive(Copy, Clone, Debug)]
+pub struct Version {
+    /// The year of this release.
+    pub year: &'static str,
+    /// The month of this release.
+    pub month: &'static str,
+    /// The patch is a positive number incremented for every compatible release on top of the major.month release.
+    pub patch: &'static str,
+    /// The commit sha for this release.
+    pub sha: &'static str,
+    /// The full version string.
+    /// Note that this extra field is used only for the uptime_metric function which requries a
+    /// &'static str.
+    pub full: &'static str,
+}
+
+impl Version {
+    /// Use for testing when you need the Version obj and a year.month &str
+    pub fn for_testing() -> Self {
+        Self {
+            year: env!("CARGO_PKG_VERSION_MAJOR"),
+            month: env!("CARGO_PKG_VERSION_MINOR"),
+            patch: env!("CARGO_PKG_VERSION_PATCH"),
+            sha: "unknown",
+            // note that this full field is needed for metrics but not for testing
+            full: const_str::concat!(
+                env!("CARGO_PKG_VERSION_MAJOR"),
+                ".",
+                env!("CARGO_PKG_VERSION_MINOR"),
+                ".",
+                env!("CARGO_PKG_VERSION_PATCH"),
+                "-unknown"
+            ),
+        }
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}.{}.{}-{}",
+            self.year, self.month, self.patch, self.sha
+        )
```

Comment:
nit: use the `full` representation here as well?

```suggestion
        write!(f, "{}", self.full)
```

Reply:
Although incidentally, I looked up why the version parameter was a static string to `uptime_metric`, and I can't find a good reason for it -- I think you could just make it a `&str` to avoid the `full` field and rely on the `Display` impl instead (even for passing the version to `uptime_metric`).

The reason I say this is that the process and chain identifier parameters are *not* `'static`, and they all get put into the same array.

Reply:
I'll have a look at this in another PR.